File: src/main/java/org/cubexmc/metro/physics/KinematicLeadSpeedPlanner.java
package org.cubexmc.metro.physics;

import java.util.List;

import org.cubexmc.metro.util.LocationUtil;

final class KinematicLeadSpeedPlanner {

    private KinematicLeadSpeedPlanner() {
    }

    static double planTargetSpeed(LocationUtil.RailType currentRailType, double baseSpeed, boolean safeMode,
            boolean poweredAscendingRailPowered, List<LocationUtil.RailType> lookaheadRailTypes) {
        double targetSpeed = LocationUtil.getSafeSpeedForRail(currentRailType, baseSpeed, safeMode);
        double minSpeedAhead = targetSpeed;

        if (safeMode && lookaheadRailTypes != null && !lookaheadRailTypes.isEmpty()) {
            int lookahead = lookaheadRailTypes.size();
            for (int i = 0; i < lookahead; i++) {
                LocationUtil.RailType aheadType = lookaheadRailTypes.get(i);
                double aheadSafe = LocationUtil.getSafeSpeedForRail(aheadType, baseSpeed, true);
                minSpeedAhead = Math.min(minSpeedAhead, aheadSafe);
                if (aheadSafe < targetSpeed - 1.0e-4) {
                    double blendFactor = (double) (i + 1) / (lookahead + 1);
                    double blended = targetSpeed * blendFactor + aheadSafe * (1.0 - blendFactor);
                    targetSpeed = Math.min(targetSpeed, blended);
                }
            }
        }

        targetSpeed = KinematicSpacingMath.applyTerrainBoost(
                poweredAscendingRailPowered,
                currentRailType,
                targetSpeed,
                baseSpeed);

        if (currentRailType == LocationUtil.RailType.CURVE && minSpeedAhead < targetSpeed) {
            targetSpeed = minSpeedAhead;
        }

        return Math.max(0.05, targetSpeed);
    }
}

File: src/main/java/org/cubexmc/metro/physics/KinematicLeadSpeedPlanner.kt
package org.cubexmc.metro.physics

import org.cubexmc.metro.util.LocationUtil
import kotlin.math.max
import kotlin.math.min

internal object KinematicLeadSpeedPlanner {

    @JvmStatic
    fun planTargetSpeed(
        currentRailType: LocationUtil.RailType,
        baseSpeed: Double,
        safeMode: Boolean,
        poweredAscendingRailPowered: Boolean,
        lookaheadRailTypes: List<LocationUtil.RailType>?,
    ): Double {
        var targetSpeed = LocationUtil.getSafeSpeedForRail(currentRailType, baseSpeed, safeMode)
        var minSpeedAhead = targetSpeed

        if (safeMode && !lookaheadRailTypes.isNullOrEmpty()) {
            val lookahead = lookaheadRailTypes.size
            for (index in 0 until lookahead) {
                val aheadType = lookaheadRailTypes[index]
                val aheadSafe = LocationUtil.getSafeSpeedForRail(aheadType, baseSpeed, true)
                minSpeedAhead = min(minSpeedAhead, aheadSafe)
                if (aheadSafe < targetSpeed - 1.0e-4) {
                    val blendFactor = (index + 1).toDouble() / (lookahead + 1)
                    val blended = targetSpeed * blendFactor + aheadSafe * (1.0 - blendFactor)
                    targetSpeed = min(targetSpeed, blended)
                }
            }
        }

        targetSpeed = KinematicSpacingMath.applyTerrainBoost(
            poweredAscendingRailPowered,
            currentRailType,
            targetSpeed,
            baseSpeed,
        )

        if (currentRailType == LocationUtil.RailType.CURVE && minSpeedAhead < targetSpeed) {
            targetSpeed = minSpeedAhead
        }

        return max(0.05, targetSpeed)
    }
}
